sway-ast/src/attribute.rs

```diff
@@ -52,10 +52,26 @@ pub enum AttributeHashKind {
     Outer(HashToken),
 }
 
+#[derive(Clone, Debug)]
+pub struct AttributeArg {
+    pub name: Ident,
+    pub value: Option<Literal>,
+}
+
+impl Spanned for AttributeArg {
+    fn span(&self) -> Span {
+        if let Some(value) = &self.value {
+            Span::join(self.name.span(), value.span())
+        } else {
+            self.name.span()
+        }
+    }
+}
+
 #[derive(Clone, Debug)]
 pub struct Attribute {
     pub name: Ident,
-    pub args: Option<Parens<Punctuated<Ident, CommaToken>>>,
+    pub args: Option<Parens<Punctuated<AttributeArg, CommaToken>>>,
 }
```

Comment:
Perhaps we should consider making `Attribute` a tree, as we'll likely end up wanting more than just one or two depths of nesting. I'd imagine it might also simplify the implementation a little. Briefly mentioned [here](https://github.com/FuelLabs/sway/issues/3576#issuecomment-1352863930).

I'm also not so sure that supporting the `=` operator like Rust does is really necessary when we can just use nesting/args for these cases too (also touched-on in the aforementioned comment).

Reply:
I was kinda thinking the same thing when looking at the code but then thought, nah this is fine.  But after looking at those examples in #3544 you linked, I agree.  And we don't really need both types.

Reply:
AFAIK Rust doesn't even parse the args and `cfg(a b c)` is like "calling" `cfg` with a TokenStream containing `a b c`. (There's support for `cfg = a b c` but that's not what we are doing here.)

If we wanted to go full Rust it would be something like:
```rs
struct Attribute {
    path: Path, // Not a name, but a proper path
    tokens: TokenStream, // No args, but just raw tokens
}
```

We were already parsing these tokens as args so I just went with it, but parsing them as whole attributes with nesting takes us a lot further away from what Rust does.

For example this:
```rs
#[cfg(target = "evm", feature = "test")]
// ...
```

becomes this:
```rs
#[cfg(target("evm"), feature("test"))]
// ...
```

which makes args of `cfg` attributes and `target` and `feature` become paths which sounds weird.

Reply:
> AFAIK Rust doesn't even parse the args

> If we wanted to go full Rust it would be something like:

> but parsing them as whole attributes with nesting takes us a lot further away from what Rust does.

My suggestion is an attempt to come up with something simpler than Rust's approach.

> For example this:
> 
> #[cfg(target = "evm", feature = "test")]
> // ...
> 
> becomes this:
> 
> #[cfg(target("evm"), feature("test"))]
> // ...

IMO they're equally as readable, but with the latter we get a simpler macro model.

> which makes args of cfg attributes and target and feature become paths which sounds weird.

Fwiw I'm not convinced paths are actually that useful for Rust macros either :sweat_smile:  I know they're most popularly used by clippy and rustfmt, but they could probably also just use nesting rather than complicate the macro model further too. All that said, even if we did have a tree of paths, I personally don't think `target` and `feature` being paths under the hood makes this example any stranger or less readable.

I decidedly didn't block this as I know our policy is to follow Rust by default unless we can come up with something unambiguously nicer. This is just one of those areas where I think Rust is a bit unnecessarily messy, but if you think the assignment syntax adds value I don't mind if we include it!